sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -214,9 +214,20 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
         .filterNot(partitionColumns.contains)
 
       // Metadata attributes are part of a column of type struct up to this point. Here we extract
-      // this column from the schema.
+      // this column from the schema and specify a matcher for that.
+      object MetadataStructColumn {
+        def unapply(attributeReference: AttributeReference): Option[AttributeReference] = {
```

Comment:
how about returning `Option[StructType]`? I'm looking at `metadataStructOpt`, and seems we only need its struct type. The same to rebinding predicates

Reply:
Unfortunately we require the exprId towards the end of the function